File: agenkit-cs/src/Agenkit/Patterns/ReflectionAgent.cs
using Agenkit.Adapters;
using Agenkit.Core;

namespace Agenkit.Patterns;

/// <summary>Configuration for ReflectionAgent.</summary>
public record ReflectionAgentConfig(
    ILlmClient LlmClient,
    int MaxReflections = 2,
    string? SystemPrompt = null,
    string? CritiquePrompt = null);

/// <summary>
/// An agent that generates a response, critiques it, and refines iteratively.
/// </summary>
public class ReflectionAgent : IAgent
{
    private readonly ILlmClient _llm;
    private readonly int _maxReflections;
    private readonly string _systemPrompt;
    private readonly string _critiquePrompt;
    private int _reflectionCount;

    /// <inheritdoc />
    public string Name => "ReflectionAgent";

    /// <inheritdoc />
    public IReadOnlyList<string> Capabilities => new[] { "reflection", "self-critique", "refinement" };

    /// <summary>Creates a new ReflectionAgent.</summary>
    public ReflectionAgent(ReflectionAgentConfig config)
    {
        ArgumentNullException.ThrowIfNull(config);
        _llm = config.LlmClient;
        _maxReflections = config.MaxReflections >= 0 ? config.MaxReflections : 2;
        _systemPrompt = config.SystemPrompt ??
            "You are a helpful assistant. Provide thorough, accurate responses.";
        _critiquePrompt = config.CritiquePrompt ??
            "Review the response above. Identify any errors, gaps, or improvements needed. Be specific and constructive.";
    }

    /// <inheritdoc />
    public async Task<Message> ProcessAsync(Message message, CancellationToken ct = default)
    {
        _reflectionCount = 0;
        var history = new List<Message>
        {
            Message.NewMessage("system", _systemPrompt),
            message
        };

        // Generate initial response
        var response = await _llm.ChatAsync(history, ct).ConfigureAwait(false);
        history.Add(response);

        // Iterative reflection
        for (int i = 0; i < _maxReflections; i++)
        {
            ct.ThrowIfCancellationRequested();

            // Ask for critique
            history.Add(Message.NewMessage("user", _critiquePrompt));
            var critique = await _llm.ChatAsync(history, ct).ConfigureAwait(false);
            history.Add(critique);

            // Ask for refined response
            history.Add(Message.NewMessage("user",
                "Based on the critique above, provide an improved response to the original question."));
            response = await _llm.ChatAsync(history, ct).ConfigureAwait(false);
            history.Add(response);
            _reflectionCount++;
        }

        return response.WithMetadata("reflection_count", _reflectionCount);
    }

    /// <inheritdoc />
    public IntrospectionResult Introspect() => new(
        Name,
        Capabilities,
        State: new Dictionary<string, object> { ["last_reflection_count"] = _reflectionCount });
}

File: agenkit-java/src/main/java/io/agenkit/patterns/ReflectionAgent.java
package io.agenkit.patterns;

import io.agenkit.adapters.LlmClient;
import io.agenkit.core.Agent;
import io.agenkit.core.IntrospectionResult;
import io.agenkit.core.Message;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * An agent that generates a response, critiques it, and refines based on feedback.
 */
public final class ReflectionAgent implements Agent {

    private final String name;
    private final LlmClient llmClient;
    private final int maxIterations;

    public ReflectionAgent(String name, LlmClient llmClient, int maxIterations) {
        this.name = name;
        this.llmClient = llmClient;
        this.maxIterations = maxIterations;
    }

    public ReflectionAgent(String name, LlmClient llmClient) {
        this(name, llmClient, 2);
    }

    @Override
    public String getName() { return name; }

    @Override
    public List<String> getCapabilities() {
        return List.of("reflection", "self_critique", "refinement");
    }

    @Override
    public CompletableFuture<Message> process(Message message) {
        return generateInitial(message).thenCompose(initial ->
                refine(message, initial.contentString(), 0));
    }

    private CompletableFuture<Message> generateInitial(Message message) {
        return llmClient.complete(List.of(
                Message.of("system", "You are a helpful assistant. Provide a thorough response."),
                message));
    }

    private CompletableFuture<Message> refine(Message original, String current, int iteration) {
        if (iteration >= maxIterations) {
            return CompletableFuture.completedFuture(
                    Message.of("assistant", current)
                            .withMetadata("iterations", iteration)
                            .withMetadata("refined", iteration > 0));
        }

        return llmClient.complete(List.of(
                Message.of("system", "Critique this response and identify specific improvements."),
                Message.of("user", "Original question: " + original.contentString()
                        + "\n\nResponse to critique:\n" + current)))
                .thenCompose(critique ->
                        llmClient.complete(List.of(
                                Message.of("system", "Improve your response based on the critique."),
                                Message.of("user", "Original question: " + original.contentString()),
                                Message.of("assistant", current),
                                Message.of("user", "Critique: " + critique.contentString()
                                        + "\n\nPlease provide an improved response.")))
                                .thenCompose(improved ->
                                        refine(original, improved.contentString(), iteration + 1)));
    }

    @Override
    public IntrospectionResult introspect() {
        Map<String, Object> state = new HashMap<>();
        state.put("maxIterations", maxIterations);
        return new IntrospectionResult(name, getCapabilities(), null, state, null);
    }
}
